### ISSUE TITLE
LUT count and Area Table

### ISSUE THREAD

**Issue by SampadChowdhury** [alanbacellar/DWN#3]
Hi,
I would like to request clarification regarding the results table you referenced. My understanding is that the software training phase utilizes 6-input LUTs. Could you please confirm whether, during hardware synthesis, all functions are mapped to only LUT6 as well, or were smaller LUT's were used too?

For example, in an MNIST configuration with a 2000-1000 setup, the reported LUT count is 4600. Does this imply that 3000 of these are LUT6s dedicated to the neural network structure, with the remaining 1600 allocated to the adder circuits?

**Comment by ZSusskind** [comment]
The hardware LUT counts in the paper are the number of physical LUT6s that Vivado reported were used post-implementation. Xilinx FPGAs can also split a single LUT6 into two LUT5s with identical inputs, and my understanding is that this will be reported as a single LUT6.

The number of hardware LUT6s used for the neural network structure is roughly equal to the number of LUT6s in the model, though sometimes Vivado is able to optimize a few of them out. The remaining LUTs are used for input deserialization and decompression, popcount / adder circuits, and the argmax circuit.

**Comment by SampadChowdhury** [comment]
That answers my question. Thankyou for the detailed explanation.

**Comment by SampadChowdhury** [comment]
Hi, reopening the issue to ask a few more questions regarding the timing values.
Are the latency, throughput, and Fmax values mentioned in the paper taken into consideration using the input deserialization and decompression units or is it just the DUT(LUT network) timing values?

**Comment by ZSusskind** [comment]
The values in the main paper are for end-to-end inference, including deserialization and decompression. The arXiv version of the paper includes results for just the model itself in Appendix D.

**Comment by johanjino** [comment]
Regarding the LUT counts and Fmax reported in the paper.

1. By Decompression do you mean the hardware implementation for thermometer encoding? 

2. In the reported results in Table 2 were the reduction/argmax (i.e. GroupSum) done using popcount or using learned reduction? I wasn't able to perform the learned reduction and achieve the same accuracy for MNIST, the paper misses these details when mentioning model hyperparameters. The reason I ask is, a pipelined tree adder implementation for GroupSum matches your Fmax, but latency and LUT count are significantly more, which indicate the GroupSum isnt mapped that way.

Thanks!

**Comment by SampadChowdhury** [comment]
for mnist, popcounts(pipelined adder tree) has been used during implementation. learned reduction is used only for tiny classifier datasets

**Comment by ZSusskind** [comment]
Data is moved onto the FPGA in a binary format since thermometer encoding is very inefficient from a compression standpoint; for instance, the unary 0011111 could be equivalently represented as the binary 101. "Decompression" refers to undoing this in order to restore the thermometer encoding for LUT inputs.

The popcount implementation is a bit more optimized than a pipelined adder tree; in particular, we use the 36b popcount trick explained here: https://fpga.org/2014/09/05/quick-fpga-hacks-population-count/

**Comment by SampadChowdhury** [comment]
Another question regarding LUT distribution.
During implementation in FPGA, did all the LUT nodes got mapped to LUT6 functions? Or did the tool break them into smaller LUT functions (LUT3, LUT4) ?
If my understanding is correct, since the train part is done using LUT6 functions, in hardware implementation too all functions should map to a LUT6 function.

**Comment by johanjino** [comment]
I observe that all the LUTLayers get almost always get exactly mapped to LUT6s. That is for the MNIST model, total 3000 LUT6 (2000+1000) are used. Rest are coming from the popcount, which may be broken down to smaller LUTs.

**Comment by SampadChowdhury** [comment]
Thanks Johan.

**Comment by SampadChowdhury** [comment]
Hi,
In the arxiv version in Table7 you've mentioned that the results are based on 250MHz, the maximum frequency of the board.
The Z 7045 fpga (from their datasheet) mentions a maximum operating frequency of  [667 MHz (-1); 800 MHz (-2); 1GHz (-3)] 

https://docs.amd.com/v/u/en-US/ds190-Zynq-7000-Overview

So is the 250 MHz number based on the LUT design or a constraint from the fpga board?
